File: src/python/3_StructuralPatterns/7_Proxy/exercise.py
from unittest import TestCase


class Person:
  def __init__(self, age):
    self.age = age

  def drink(self):
    return 'drinking'

  def drive(self):
    return 'driving'

  def drink_and_drive(self):
    return 'driving while drunk'

class ResponsiblePerson:
  def __init__(self, person):
    self.person = person

  @property
  def age(self):
    return self.person.age

  @age.setter
  def age(self, value):
    self.person.age = value

  def drink(self):
    if self.age >= 18:
      return self.person.drink()
    return 'too young'

  def drive(self):
    if self.age >= 16:
      return self.person.drive()
    return 'too young'

  def drink_and_drive(self):
    return 'dead'

class Evaluate(TestCase):
  def test_exercise(self):
    p = Person(10)
    rp = ResponsiblePerson(p)

    self.assertEqual('too young', rp.drive())
    self.assertEqual('too young', rp.drink())
    self.assertEqual('dead', rp.drink_and_drive())

    rp.age = 20

    self.assertEqual('driving', rp.drive())
    self.assertEqual('drinking', rp.drink())
    self.assertEqual('dead', rp.drink_and_drive())

File: src/swift/3_StructuralPatterns/7_Proxy/Exercise.swift
import Foundation
import XCTest

class Person
{
  var age: Int = 0

  func drink() -> String
  {
    return "drinking"
  }

  func drive() -> String
  {
    return "driving"
  }

  func drinkAndDrive() -> String
  {
    return "driving while drunk"
  }
}

class ResponsiblePerson
{
  private let person: Person

  init(person: Person)
  {
    self.person = person
  }

  var age: Int
  {
    get { return person.age }
    set(value) { person.age = value }
  }

  func drink() -> String
  {
    return (age >= 18) ? person.drink() : "too young"
  }

  func drive() -> String
  {
    return (age >= 16) ? person.drive() : "too young"
  }

  func drinkAndDrive() -> String
  {
    return "dead"
  }
}

class UMBaseTestCase : XCTestCase {}

//@testable import Test

class Evaluate: UMBaseTestCase
{
  func simpleTest()
  {
    let p = Person()
    p.age = 10
    let rp = ResponsiblePerson(person: p)

    XCTAssertEqual("too young", rp.drive(), "Should be too young to drive at age 10")
    XCTAssertEqual("too young", rp.drink(), "Should be too young to drink vodka at age 10")
    XCTAssertEqual("dead", rp.drinkAndDrive(), "Drinking while driving should lead to certain death")

    rp.age = 20

    XCTAssertEqual("driving", rp.drive(), "Driving at age 20 should be OK")
    XCTAssertEqual("drinking", rp.drink(), "Drinking at age 20 should be OK so long as you do it in moderation")
    XCTAssertEqual("dead", rp.drinkAndDrive(), "Driving while drunk leads to hell")
  }
}

extension Evaluate
{
  static var allTests : [(String, (Evaluate) -> () throws -> Void)]
  {
    return [
      ("simpleTest", simpleTest)
    ]
  }
}

func main()
{
  XCTMain([testCase(Evaluate.allTests)])
}

main()